### PR TITLE
Development

### PR DESCRIPTION
**Pull request by stuttste**
Fleshed out Haunts and Landmarks in this one. Decided to hold off on Delves for now since our session 1 is tomorrow and I don't wanna accidentally break something before then. lol

Noticed you pushed 0.4.7, so merged and changed mine to 0.4.8

### CODE REVIEW

**Comment by stuttste** [comment]
Left out Default Stress on the Landmarks. I'll create an issue on it so we don't forget.

File: src/items/haunt/template.json [hunk 1]
```diff
@@ -1,7 +1,9 @@
 {
     "Item": {   
         "haunt": {
-            "templates": ["has_description"]
+            "die_sizes": ["d4", "d6", "d8", "d10", "d12"],
```
Comment (hitcherland) [inline src/items/haunt/template.json:4]: We don't need to add die_sizes to an item template; these templates are for per-item properties and die sizes are global.

File: src/items/haunt/sheet.js [hunk 2]
```diff
@@ -1,12 +1,128 @@
...
+
+            roll.toMessage({
+                flavor: `${localizeHeart(this.item.name)} (<span class="item-type">${item.type}</span>)`,
+                speaker: {alias: "GM"}
+            });
+        });
+
+        html.find('[data-action=upgrade]').click(async ev => {
+            const target = $(ev.currentTarget);
+            const dieSizes = this.item.data.data.die_sizes;
```
Comment (hitcherland) [inline src/items/haunt/sheet.js:75]: This should use `game.heart.die_sizes` instead.

File: src/items/haunt/sheet.js [hunk 3]
```diff
@@ -1,12 +1,128 @@
...
+                }
+            });
+
+            await this.item.updateChildren(childrenUpdates);
+            this.item.update(updates);
+        });
+
+        html.find('[data-action=downgrade]').click(async ev => {
+            const target = $(ev.currentTarget);
+            const dieSizes = this.item.data.data.die_sizes;
```
Comment (hitcherland) [inline src/items/haunt/sheet.js:96]: This should use `game.heart.die_sizes` instead.

File: src/items/haunt/lang/en.json [hunk 4]
```diff
@@ -1,6 +1,15 @@
 {
     "haunt": {
         "label-single": "Haunt",
-        "label-multiple": "Haunts"
+        "label-multiple": "Haunts",
+        "upgrade": "Upgrade",
+        "downgrade": "Downgrade"
+    },
+    "service": {
```
Comment (hitcherland) [inline src/items/haunt/lang/en.json:8]: If we're keeping services, this belongs in the "service/lang/{xx}.json" files

File: src/items/haunt/lang/es.json [hunk 5]
```diff
@@ -1,6 +1,15 @@
 {
     "haunt": {
         "label-single": "Tormento",
-        "label-multiple": "Tormentos"
+        "label-multiple": "Tormentos",
+        "upgrade": "Mejora",
+        "downgrade": "Degradar"
+    },
+    "service": {
```
Comment (hitcherland) [inline src/items/haunt/lang/es.json:8]: As above, this would belong in service/lang/es.json

File: src/items/haunt/preview.html [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+            {{localizeHeart "upgradeTrack" "label-single"}}
+            <div class="right">
+                <a data-action="upgrade" data-type="haunt"><span class="fas fa-level-up-alt"></span>{{localizeHeart "haunt" "upgrade"}}</a>
+                <a data-action="downgrade" data-type="haunt"><span class="fas fa-level-down-alt"></span>{{localizeHeart "haunt" "downgrade"}}</a>
+            </div>
+        </h2>
+        <div class="red" data-target='data.upgradeTrack'>{{{ordered-checkable item.data.data.upgradeTrack 6}}}</div>
+    </div>
+
+    <div class="section {{localizeHeart service 'label-multiple'}}">
```
Comment (hitcherland) [inline src/items/haunt/preview.html:16]: We don't need the distinct item "service" here to represent different resistance rolls, that capability should be directly added to Haunts.

File: src/items/haunt/template.json [hunk 7]
```diff
@@ -1,7 +1,9 @@
...
+            "templates": ["has_description", "has_children", "is_activatable"],
```
Comment (hitcherland) [inline src/items/haunt/template.json:5]: As discussed elsewhere, haunts shouldn't need children.

File: src/items/haunt/template.json [hunk 8]
```diff
@@ -1,7 +1,9 @@
...
+            "upgradeTrack": ""
         }
```
Comment (hitcherland) [inline src/items/haunt/template.json:7]: Let's add a `resistances: []` section here, which should replace the need for the `services` item.

File: src/items/haunt/template.json (hunk 1 above)
Comment (stuttste) [reply src/items/haunt/template.json:4]: I should be able to get this removed. Had trouble getting to the original die_sizes parameter but the below should give me what I need there I think.

File: src/items/haunt/sheet.js (hunk 2 above)
Comment (stuttste) [reply src/items/haunt/sheet.js:75]: Got it. I should be able to swap to that.

File: src/items/haunt/sheet.js (hunk 3 above)
Comment (stuttste) [reply src/items/haunt/sheet.js:96]: As above

File: src/items/haunt/lang/en.json (hunk 4 above)
Comment (stuttste) [reply src/items/haunt/lang/en.json:8]: Makes sense. Think this was leftover from my initial attempts to keep it all contained within haunts. Will correct once we settle which path we wanna take.

File: src/items/haunt/lang/es.json (hunk 5 above)
Comment (stuttste) [reply src/items/haunt/lang/es.json:8]: As above

**Comment by stuttste** [comment]
For the life of me, though, I cannot get selectOptions to work correctly for the Haunt object. It seems to be related to me extending the base object, but I can't figure out what I'm doing differently to cause the options list to be undefined. Attempting to just plug and play the `{{selectOptions die_sizes selected=<resistances array>}}` just will not work.

**Comment by stuttste** [comment]
image

This is the result if we decide to omit the service-level object. Is this the path we want to take? (Ignore the middle "Services" section.)

**Comment by stuttste** [comment]
Perfect. Yep, I'll be sure that's in there as well. Wanted to clear the format with you before continuing. Thanks. Will get another commit posted soonish.

**Comment by stuttste** [comment]
Ok, committed and reviewed. I think I got everything we mentioned.

File: release-notes/v0.4.8/lang/en.json [hunk 9]
```diff
@@ -4,7 +4,7 @@
             "version": "v0.4.8",
             "changes":[
                 "Another update from stuttste",
-                "Created 'Service' item type to represent services rendered by haunts",
+                "Created 'Service' item type to represent services rendered by haunts; EDIT: Service object removed in favor of internal array/object of Haunts.",
```
Comment (hitcherland) [inline release-notes/v0.4.8/lang/en.json:7]: Users won't knowthe context for this EDIT, please remove.

File: src/items/haunt/sheet.js [hunk 10]
```diff
@@ -57,55 +57,71 @@ export default class extends HeartItemSheet {
...
+            const id = target.closest ('[data-id]').data('id');
+            const service = this.item.data.data.resistances[id];
+            const item = {data:{data:{die_size:service.die_size}}};
 
             const roll = game.heart.rolls.ItemRoll.build({item});
             await roll.evaluate({async: true});
 
             roll.toMessage({
-                flavor: `${localizeHeart(this.item.name)} (<span class="item-type">${item.type}</span>)`,
+                flavor: `${localizeHeart(this.item.name)} (<span class="item-type">${this.item.type}</span>)`,
```
Comment (hitcherland) [inline src/items/haunt/sheet.js:70]: We should include the resistance type in the roll output, i.e. add `localizeHeart(service.resistance)` in there somewhere

**Review by hitcherland — CHANGES_REQUESTED** [review]
This is just about ready to be submitted! I noticed some problems dragging Haunts out of Landmarks and into the items tab, but I didn't have time to check if that was a general problem or not, so it can wait for a future release.

**Comment by stuttste** [comment]
Great! Changes made. Yeah, drag-n-drop is something I wanted to look into at some point too. Mainly, seems like dragging and dropping resources from adversary to characters didn't seem to work when I tried it. Would be a convenient thing to have when DM'ing. Also a 'share'/'send' button for players to give resources to one another.

**Comment by stuttste** [comment]
You're absolutely right. I hadn't even considered that. Thanks! Should be updated now.

**Pull request merged**